app/controllers/users/piv_cac_authentication_setup_controller.rb

```diff
@@ -0,0 +1,57 @@
+module Users
+  class PivCacAuthenticationSetupController < ApplicationController
+    include UserAuthenticator
+
+    before_action :confirm_two_factor_authenticated
+    before_action :authorize_piv_cac_setup, only: :new
+    before_action :authorize_piv_cac_disable, only: :disable
+
+    def new
+      analytics.track_event(Analytics::USER_REGISTRATION_PIV_CAC_SETUP_VISIT)
+      if params.key?(:token)
+        process_token(params[:token])
+      else
+        render :new
+      end
+    end
+
+    def disable
+      current_user.update!(x509_dn_uuid: nil)
+      Event.create(user_id: current_user.id, event_type: :piv_cac_disabled)
+      flash[:success] = t('notices.piv_cac_disabled')
+      redirect_to account_url
+    end
+
+    private
+
+    def process_token(token)
+      @user_piv_cac_form = UserPivCacSetupForm.new(current_user, token)
+      result = @user_piv_cac_form.submit
+      if result.success?
+        process_valid_submission
+      else
+        process_invalid_submission
+      end
+    end
+
+    def process_valid_submission
+      flash[:success] = t('notices.piv_cac_configured')
+      # TODO: make this redirect to the next step in the process if they are signing up
+      # for the first time.
+      redirect_to account_url
+    end
+
+    def process_invalid_submission
+      @presenter = PivCacAuthenticationSetupPresenter.new(@user_piv_cac_form)
+      render :error
+    end
+
+    def authorize_piv_cac_disable
+      redirect_to account_url if user_fully_authenticated? && !current_user.piv_cac_enabled?
```

Comment:
The `user_fully_authenticated?` condition is not necessary here since we already have the `before_action :confirm_two_factor_authenticated` at the top.